File: backend/RescueSriLanka.Api/Models/SafetyZone.cs
using System.ComponentModel.DataAnnotations;

namespace RescueSriLanka.Api.Models;

/// <summary>
/// A circular area with a safety classification. Most zones are recomputed from
/// active incidents; coordinators can also declare one by hand, and those are
/// never overwritten by the recompute.
/// </summary>
public class SafetyZone
{
    public Guid Id { get; set; } = Guid.NewGuid();

    [MaxLength(200)]
    public required string Name { get; set; }

    public ZoneStatus Status { get; set; } = ZoneStatus.Caution;

    public ZoneSource Source { get; set; } = ZoneSource.DerivedFromIncident;

    public double CenterLatitude { get; set; }
    public double CenterLongitude { get; set; }
    public int RadiusMeters { get; set; }

    [MaxLength(100)]
    public string? District { get; set; }

    [MaxLength(500)]
    public string? Rationale { get; set; }

    /// <summary>The incident this zone was derived from, when it was not declared manually.</summary>
    public Guid? SourceIncidentId { get; set; }
    public Incident? SourceIncident { get; set; }

    public bool IsActive { get; set; } = true;

    public DateTime ComputedAt { get; set; } = DateTime.UtcNow;
    public DateTime? ExpiresAt { get; set; }
}

File: mobile/lib/models/safety_zone.dart
/// Mirrors SafetyZoneDto from the API.
class SafetyZone {
  const SafetyZone({
    required this.id,
    required this.name,
    required this.status,
    required this.source,
    required this.centerLatitude,
    required this.centerLongitude,
    required this.radiusMeters,
    this.district,
    this.rationale,
    this.sourceIncidentId,
  });

  final String id;
  final String name;
  final String status;
  final String source;
  final double centerLatitude;
  final double centerLongitude;
  final int radiusMeters;
  final String? district;
  final String? rationale;
  final String? sourceIncidentId;

  bool get isManual => source == 'ManualOverride';

  factory SafetyZone.fromJson(Map<String, dynamic> json) => SafetyZone(
        id: json['id'] as String,
        name: json['name'] as String? ?? 'Zone',
        status: json['status'] as String? ?? 'Caution',
        source: json['source'] as String? ?? 'DerivedFromIncident',
        centerLatitude: (json['centerLatitude'] as num).toDouble(),
        centerLongitude: (json['centerLongitude'] as num).toDouble(),
        radiusMeters: (json['radiusMeters'] as num?)?.toInt() ?? 1000,
        district: json['district'] as String?,
        rationale: json['rationale'] as String?,
        sourceIncidentId: json['sourceIncidentId'] as String?,
      );
}

/// Mirrors ZoneCheckResultDto — the answer to "is this spot safe?".
class ZoneCheck {
  const ZoneCheck({
    required this.status,
    required this.message,
    required this.nearbyIncidentCount,
  });

  final String status;
  final String message;
  final int nearbyIncidentCount;

  factory ZoneCheck.fromJson(Map<String, dynamic> json) => ZoneCheck(
        status: json['status'] as String? ?? 'Safe',
        message: json['message'] as String? ?? '',
        nearbyIncidentCount: (json['nearbyIncidentCount'] as num?)?.toInt() ?? 0,
      );
}
